trace-sdk/src/test/java/io/bitrise/trace/scheduler/ExecutorSchedulerTest.java

```diff
@@ -0,0 +1,112 @@
+package io.bitrise.trace.scheduler;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+import android.content.Context;
+import androidx.annotation.Nullable;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+/**
+ * Unit tests for {@link ExecutorScheduler}.
+ */
+public class ExecutorSchedulerTest {
+
+  private final Context mockContext = Mockito.mock(Context.class);
+  private final Runnable mockRunnable = Mockito.mock(Runnable.class);
+  private final long initialDelay = 600;
+
+  @Test
+  public void scheduleDelayed_noRunnable() {
+    final ExecutorScheduler executorScheduler = new ExecutorScheduler(mockContext, null);
+
+    assertNull(executorScheduler.runnable);
+    assertNull(executorScheduler.scheduleDelayed(initialDelay));
+  }
+
+  @Test
+  public void scheduleDelayed_noExecutiveService() {
+    final ExecutorScheduler executorScheduler = new ExecutorScheduler(mockContext, mockRunnable);
+    assertNull(executorScheduler.scheduledExecutorService);
+
+    executorScheduler.scheduleDelayed(initialDelay);
+
+    final ScheduledThreadPoolExecutor executor = getScheduledThreadPoolExecutorFromService(
+        executorScheduler.scheduledExecutorService);
+    assertEquals(1, executor.getPoolSize());
+  }
+
+  @Test
+  public void scheduleDelayed_intervalGreaterThanZero() {
+    final ExecutorScheduler executorScheduler =
+        new ExecutorScheduler(mockContext, mockRunnable, initialDelay, 200);
+
+    executorScheduler.scheduleDelayed(initialDelay);
+
+    assertNotNull(executorScheduler.scheduledFuture);
+    assertTrue(executorScheduler.scheduledFuture.getDelay(TimeUnit.MILLISECONDS)
+        > (initialDelay - 100));
+    // we cannot get the delay immediately, several milliseconds will have passed.
```

Comment:
Not sure how you feel about this - we can't get the exact time that was set because the time will have elapsed - i figured if it's within 100ms of what we set it's probably okay. If it's instant the time will be zero or negative (the delay has passed) so it's easy to compare that? What do you think?

Reply:
I have no strong opinion on this.